staging/src/k8s.io/apiserver/pkg/server/filters/longrunning.go

```diff
@@ -39,3 +40,83 @@ func BasicLongRunningRequestCheck(longRunningVerbs, longRunningSubresources sets
 		return false
 	}
 }
+
+// WithLongRunningRequestTermination starts closing long running requests upon receiving TerminationStartCh signal.
+//
+// It does it by running two additional go routines.
+// One for running the request and the second one for intercepting the termination signal and propagating it to the requests.
+// If the request is not terminated after receiving a signal from inFlightRequestsFinished it will be forcefully killed.
+//
+// This filter exists because sometimes propagating the termination signal is not enough.
+// It turned out that long running requests might block on:
+//  io.Read() for example in https://golang.org/src/net/http/httputil/reverseproxy.go
+//  http2.(*serverConn).writeDataFromHandler which might be actually an issue with the std lib itself
+//
+// Instead of trying to identify current and future issues we provide a filter that ensures terminating long running requests.
+//
+// Also note that upon receiving termination signal the http server sends
+// sends GOAWAY with ErrCodeNo to tell the client we're gracefully shutting down.
+// But the connection isn't closed until all current streams are done.
+func WithLongRunningRequestTermination(handler http.Handler, longRunningFunc apirequest.LongRunningRequestCheck, terminationStartCh <-chan struct{}, inFlightRequestsFinished <-chan struct{}) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		requestInfo, found := apirequest.RequestInfoFrom(req.Context())
+		if !found || longRunningFunc == nil {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		isLongRunning := longRunningFunc(req, requestInfo)
+		if !isLongRunning {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		ctx, cancelCtxFn := context.WithCancel(req.Context())
+		defer cancelCtxFn()
+		req = req.WithContext(ctx)
+
+		errCh := make(chan interface{}, 2)
+		// note that is is okay to leave the errCh open
+		// eventually it will be garbage collected
+		doneCh := make(chan struct{})
+
+		go func() {
+			defer func() {
+				err := recover()
+				select {
+				case errCh <- err:
+					return
+				}
+			}()
+			select {
+			case <-terminationStartCh:
+				cancelCtxFn()
+				select {
+				case <-inFlightRequestsFinished:
+					panic(http.ErrAbortHandler)
+				case <-doneCh:
+					return
+				}
+			case <-doneCh:
+				return
+			}
+		}()
```

Comment:
can we get rid of one of the go routines?

Reply:
I don't think so, mainly because
- the handler can block indefinitely
- and we need to propagate panic from the other goroutine